core/services/relay/evm/evm.go

```diff
@@ -305,23 +315,75 @@ func (r *Relayer) NewMedianProvider(rargs relaytypes.RelayArgs, pargs relaytypes
 	if err != nil {
 		return nil, err
 	}
-	contractTransmitter, err := newContractTransmitter(r.lggr, rargs, pargs.TransmitterID, configWatcher)
+
+	var relayConfig RelayConfig
+	if err = json.Unmarshal(rargs.RelayConfig, &relayConfig); err != nil {
+		return nil, err
+	}
+	var contractTransmitter ocrtypes.ContractTransmitter
+	var reportCodec median.ReportCodec
+	if relayConfig.MercuryConfig != nil {
+		r.lggr.Debugf("Mercury mode enabled for job %d", rargs.JobID)
+		contractTransmitter, reportCodec, err = r.NewMercuryMedianProvider(relayConfig)
+	} else {
+		r.lggr.Debugf("On-chain mode enabled for job %d", rargs.JobID)
+		reportCodec = evmreportcodec.ReportCodec{}
+		contractTransmitter, err = newContractTransmitter(r.lggr, rargs, pargs.TransmitterID, configWatcher)
+	}
 	if err != nil {
 		return nil, err
 	}
-	medianContract, err := newMedianContract(configWatcher.contractAddress, configWatcher.chain, rargs.JobID, r.db, r.lggr)
+
+	medianContract, err := newMedianContract(configWatcher.ContractConfigTracker(), configWatcher.contractAddress, configWatcher.chain, rargs.JobID, r.db, r.lggr, relayConfig.MercuryConfig != nil)
 	if err != nil {
 		return nil, err
 	}
 	return &medianProvider{
 		configWatcher:       configWatcher,
-		reportCodec:         evmreportcodec.ReportCodec{},
+		reportCodec:         reportCodec,
 		contractTransmitter: contractTransmitter,
 		medianContract:      medianContract,
 	}, nil
+
+}
+
+func (r *Relayer) NewMercuryMedianProvider(relayConfig RelayConfig) (contractTransmitter ocrtypes.ContractTransmitter, reportCodec median.ReportCodec, err error) {
```

Comment:
Can this be moved to the evm.mercury package, to keep mercury specific code in there?

Reply:
Sadly not :( it creates dependency cycles